src/lib/communication/DY.Lib.Communication.API.fst

```diff
@@ -0,0 +1,208 @@
+open DY.Core
+open DY.Lib.State.PKI
+open DY.Lib.State.PrivateKeys
+open DY.Lib.Event.Typed
+
+#set-options "--fuel 0 --ifuel 0 --z3cliopt 'smt.qi.eager_threshold=100'"
+
+(*** Messages ***)
+
+[@@with_bytes bytes]
+type communication_message_base = {
+  sender:principal;
+  receiver:principal;  
+  payload:bytes;
+}
+
+%splice [ps_communication_message_base] (gen_parser (`communication_message_base))
+%splice [ps_communication_message_base_is_well_formed] (gen_is_well_formed_lemma (`communication_message_base))
+
+[@@with_bytes bytes]
+type communication_message_sign = {
+  sender:principal;
+  receiver:principal;  
+  payload:bytes;
+  signature:bytes;
+}
+
+%splice [ps_communication_message_sign] (gen_parser (`communication_message_sign))
+%splice [ps_communication_message_sign_is_well_formed] (gen_is_well_formed_lemma (`communication_message_sign))
+
+[@@with_bytes bytes]
+type communication_message = 
+  | CM: msg:communication_message_base -> communication_message
+  | CMSign: msg:communication_message_sign -> communication_message
+
+#push-options "--ifuel 1 --fuel 0"
+%splice [ps_communication_message] (gen_parser (`communication_message))
+%splice [ps_communication_message_is_well_formed] (gen_is_well_formed_lemma (`communication_message))
+#pop-options
+
+instance parseable_serializeable_bytes_communication_message: parseable_serializeable bytes communication_message
+  = mk_parseable_serializeable ps_communication_message
+
+(*** Events ***)
+
+#push-options "--ifuel 1 --fuel 0"
+[@@with_bytes bytes]
+type communication_event =
+  | CommConfSendMsg: sender:principal -> receiver:principal -> payload:bytes -> communication_event
+  | CommConfReceiveMsg: sender:principal -> receiver:principal -> payload:bytes -> communication_event
+  | CommAuthSendMsg: sender:principal -> payload:bytes -> communication_event
+  | CommAuthReceiveMsg: sender:principal -> receiver:principal -> payload:bytes -> communication_event
+
+%splice [ps_communication_event] (gen_parser (`communication_event))
+%splice [ps_communication_event_is_well_formed] (gen_is_well_formed_lemma (`communication_event))
+#pop-options
+
+instance event_communication_event: event communication_event = {
+  tag = "DY.Lib.Communication.Event";
+  format = mk_parseable_serializeable ps_communication_event;
+}
+
+(*** Layer Setup ***)
+
+val comm_layer_pkenc_tag: string
+let comm_layer_pkenc_tag = "DY.Lib.Communication.PkEnc.PublicKey"
+
+val comm_layer_sign_tag: string
+let comm_layer_sign_tag = "DY.Lib.Communication.Sign.PublicKey"
+
+type communication_keys_sess_ids = {
+  pki: state_id;
+  private_keys: state_id;
+}
+
+(*** Communication Layer ***)
+
+(**** Confidential Send and Receive Functions ****)
+
+val encrypt_message: principal -> principal -> bytes -> bytes -> bytes -> bytes
+let encrypt_message sender receiver payload pk nonce =
+  let msg = CM {sender; receiver; payload} in
+  let msg_bytes = serialize communication_message msg in
+  pk_enc pk nonce msg_bytes
```

Comment:
I am thinking of only encrypting the payload. This should allow us to have stronger security guarantees in the crypto predicate because we don't have to serialize the payload twice (once on the higher layer and once on the communication layer). Furthermore, this does not change the guarantees we can get from this function because the sender is not authenticated.

Reply:
I don't understand how encrypting less things allows you to have stronger guarantees in the crypto predicate, can you expand on that? This seems a bit fishy to me because e.g. encrypting the sender / receiver is needed for the security of NSL (see Lowe attack on the NS protocol).

Reply:
By stronger security guarantees, I mean that we can write `(get_label payload) == (join (principal_label sender) (principal_label receiver)) /\` in the crypto predicates. So, I am only talking about the payload. I fear that we run into problems later if we can't say `==` for the label of the payload. Another reason for not encrypting the sender and receiver is that TLS does not hide the sender and receiver IP addresses/identities either.
For the NSL example, I would argue that if you need the sender and receiver values protected, then the application layer puts this information into the payload.

Reply:
I realized that as soon as the `payload` is serialized by comparse we cannot show `==` anymore. So, I will leave it as it is.